### PR TITLE
Merge of analysis tools feature branch to main

### PR DESCRIPTION
**Pull request by ago109**
This is a merger of the analysis-tools feature branch over to main -- this feature branch includes improvements/additions to specifically `ngclearn.utils`, particularly with the inclusion of the linear and attention probes for encoding vector downstream analysis (often done in self-supervised learning experiments).

### CODE REVIEW

File: ngclearn/components/__init__.py
Comment (rxng8) [inline ngclearn/components/__init__.py:1]: No changes in this file

File: ngclearn/utils/model_utils.py [hunk 1]
```diff
@@ -532,25 +635,44 @@ def threshold_cauchy(x, lmbda):
...
+
+        shift: the compensating mean/shift factor/parameters (to undo mean subtraction)
+
+        scale: the compensating re-scaling factor/parameters (to undo standard deviation division)
+
+    Returns:
+        layer-normalized data samples `x`
+    """
+    xmu = jnp.mean(x, axis=1, keepdims=True)
+    xsigma = jnp.sqrt(jnp.mean(jnp.square(x - xmu)).clip(min=1e-6))
```
Comment (rxng8) [inline ngclearn/utils/model_utils.py:654]: I think we should be very careful here and handle the case `sqrt(0)`, since `sqrt` does not have gradient at 0. Recommend changing to `xsigma = jnp.sqrt(jnp.mean(jnp.square(x - xmu).clip(min=1e-6)).clip(min=1e-6))`

File: ngclearn/utils/model_utils.py [hunk 2]
```diff
@@ -532,25 +635,44 @@ def threshold_cauchy(x, lmbda):
...
+    _x = (x - xmu)/(xsigma + 1e-6)
```
Comment (rxng8) [inline ngclearn/utils/model_utils.py:655]: Since we have clipped the standard deviation `xsigma`, there is no need for it to be added with another `1e-6`, I think this part is fine, it's your choice to remove it

File: ngclearn/utils/analysis/probe.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+            if dev_data is not None:
+                print_string = f"\r{ii} L = {L / Ns:.3f} Acc = {acc / Ns:.2f}  Dev.Acc = {best_acc:.2f}"
+            else:
+                print_string = f"\r{ii} L = {L / Ns:.3f} Acc = {acc / Ns:.2f}"
+
+            if hasattr(self, "eta"):
+                print_string += f"  LR = {getattr(self, 'eta'):.6f}"
+            
+            print(print_string, end = "")
```
Comment (rxng8) [inline ngclearn/utils/analysis/probe.py:160]: I added this a while ago for the probe class to reflect the learning rate, e.g., `eta` throughout time steps. Please check to see if you want this to be in the main branch

File: ngclearn/utils/model_utils.py [hunk 4]
```diff
@@ -487,11 +590,12 @@ def softmax(x, tau=0.0):
     Returns:
         a (N x D) probability distribution output block
     """
+    # TODO: Do we need to also consider for edge case the division of tau by 0 is invalid?
     if tau > 0.0:
         x = x / tau
     max_x = jnp.max(x, axis=1, keepdims=True)
     exp_x = jnp.exp(x - max_x)
-    return exp_x / jnp.sum(exp_x, axis=1, keepdims=True)
+    return exp_x / jnp.sum(exp_x, axis=1, keepdims=True) # TODO: We also have to take care of the case where the sum is 0
```
Comment (rxng8) [inline ngclearn/utils/model_utils.py:598]: Please check these issue

**Review by rxng8 — CHANGES_REQUESTED** [review]

File: ngclearn/utils/model_utils.py [hunk 5]
```diff
@@ -72,17 +72,22 @@ def create_function(fun_name, args=None):
...
         function fx, first derivative of function (w.r.t. input) dfx
     """
-    fx = None
-    dfx = None
+    fx = None ## the function
+    dfx = None ## the first derivative of function w.r.t. its input
     if fun_name == "tanh":
         fx = tanh
         dfx = d_tanh
+    elif "bkwta" in fun_name:
```
[the remark this answers is not in the document]
Comment (ago109) [reply ngclearn/utils/model_utils.py:88]: I agree, let's change it to be consistent

File: ngclearn/utils/model_utils.py (hunk 1 above)
Comment (ago109) [reply ngclearn/utils/model_utils.py:654]: Ok, I agree with this -- let's change it to the proposed line

File: ngclearn/utils/model_utils.py (hunk 2 above)
Comment (ago109) [reply ngclearn/utils/model_utils.py:655]: I think we can remove the extra `+1e-6` now that the clipping prevents bad values

File: ngclearn/utils/analysis/probe.py (hunk 3 above)
Comment (ago109) [reply ngclearn/utils/analysis/probe.py:160]: Yeah, let's keep it, since we found that scheduled/changing learning rates work better for this probe

File: ngclearn/utils/model_utils.py (hunk 4 above)
Comment (ago109) [reply ngclearn/utils/model_utils.py:598]: Ok, i'll add a small `+ 1e-6` or min-clip to prevent division by 0 in the denominator

**Review by rxng8 — APPROVED** [review]
